// @fontoxml/fontoxml-development-tools/src/modules/editor/src/config.js
export const defaultBabelPlugins = [
	// Cleanup
	'cleanupRegisterXQueryModulePlugin',

	// Convert code
	'convertAmdToEsmPlugin',
	'rewriteConfigImportsPlugin',
	'rewriteImportFileExtensionPlugin',
	'rewriteInteropDefaultPlugin',
	'rewriteJsonLoaderPlugin',
	'rewriteMainJsImportsPlugin',
	'rewriteSrcPathPlugin',
	'rewriteSvgLoaderPlugin',
];

export const defaultBabelValidationPlugins = ['validatePlugin'];

export const importAliases = [
	{
		alias: 'diff',
		devImport: 'fontoxml-vendors/src/diff',
		import: 'fontoxml-vendors/src/diff.min',
	},
	{
		alias: 'diff-match-patch',
		import: 'fontoxml-vendors/src/diff-match-patch',
	},
	{
		alias: 'fds/components',
		import: 'fontoxml-vendors/src/fds/components',
	},
	{
		alias: 'fds/lib/bundle',
		import: 'fontoxml-vendors/src/fds/lib/bundle',
	},
	{ alias: 'fds/system', import: 'fontoxml-vendors/src/fds/system' },
	{
		alias: 'fonto-xml-parser',
		import: 'fontoxml-vendors/src/fonto-xml-parser/fonto-xml-parser',
	},
	{ alias: 'fontoxpath', import: 'fontoxml-vendors/src/fontoxpath' },
	{
		alias: 'fontoxpath-cache',
		import: 'fontoxml-vendors/src/fontoxpath-cache',
	},
	{ alias: 'glamor', import: 'fontoxml-vendors/src/glamor' },
	{
		alias: 'messageformat',
		devImport: 'fontoxml-vendors/src/messageformat',
		import: 'fontoxml-vendors/src/messageformat.min',
	},
	{
		alias: 'prop-types',
		devImport: 'fontoxml-vendors/src/prop-types',
		import: 'fontoxml-vendors/src/prop-types.min',
	},
	{
		alias: 'react',
		devImport: 'fontoxml-vendors/src/react',
		import: 'fontoxml-vendors/src/react.min',
	},
	{
		alias: 'react-dnd',
		devImport: 'fontoxml-vendors/src/ReactDnD',
		import: 'fontoxml-vendors/src/ReactDnD.min',
	},
	{
		alias: 'react-dnd-html5-backend',
		devImport: 'fontoxml-vendors/src/ReactDnDHTML5Backend',
		import: 'fontoxml-vendors/src/ReactDnDHTML5Backend.min',
	},
	{
		alias: 'react-dom',
		devImport: 'fontoxml-vendors/src/react-dom',
		import: 'fontoxml-vendors/src/react-dom.min',
	},
	{
		alias: 'react-router-dom',
		devImport: 'fontoxml-vendors/src/react-router-dom',
		import: 'fontoxml-vendors/src/react-router-dom.min',
	},
	{
		alias: 'react-spring',
		import: 'fontoxml-vendors/src/react-spring',
	},
	{
		alias: 'resize-observer-polyfill',
		import: 'fontoxml-vendors/src/ResizeObserver',
	},
	{ alias: 'slimdom', import: 'fontoxml-vendors/src/slimdom' },
	{ alias: 'superagent', import: 'fontoxml-vendors/src/superagent' },
	{ alias: 'unicodejs', import: 'fontoxml-vendors/src/unicodejs' },
	{ alias: 'whynot', import: 'fontoxml-vendors/src/whynot' },
	{ alias: 'xspattern', import: 'fontoxml-vendors/src/xspattern' },
];

export const mainJsModuleNames = [
	'fontoxml-base-flow',
	'fontoxml-block-flow',
	'fontoxml-blueprints',
	'fontoxml-core',
	'fontoxml-dom-utils',
	'fontoxml-schema-experience',
	'fontoxml-sectioning-flow',
	'fontoxml-shims',
	'fontoxml-templated-views',
];

export const namedExportModuleNames = [
	'fds/components',
	'fds/system',
	'fontoxpath-cache',
	'glamor',
	'slimdom',
];
